README
NSCA
====

## Nagios Service Check Acceptor

TL;DR? You can jump straight to [Compiling](#compiling) and [Installing](#installing).

Purpose
-------

The purpose of this addon is to allow you to send service check
results to a central monitoring server running Nagios in a secure
manner.


Explanation
-----------

There are two pieces to this addon:

1. `nsca`

   This program runs as a daemon on the central server
   that runs Nagios. It listens for host and service
   check results from remote machines (sent using the
   `send_nsca` program described below). Upon receiving
   data from a remote client, the daemon will make a
   **very** basic attempt at validating the data it has
   received from the client. This is done by decrypting
   the data with the password stored in the `nsca.cfg` 
   file. If the decrypted data looks okay (i.e. it was
   originally encrypted by the `send_ncsa` program using
   the same password), the daemon will make entries in
   the Nagios external command file telling Nagios
   to process the host or service check result.

   Notes: 

   1. The nsca daemon must have sufficient rights
      to open the Nagios command file for writing.  

   2. Also, Nagios will only process passive service check
      results that it finds in the external command file
      if the service has been defined in the host config
      file (i.e. `hosts.cfg`) and it is being monitored.


2. `send_nsca`

   This is the client program that is used to send
   service check information from a remote machine to
   the nsca daemon on the central machine that runs
   Nagios. Service check information is read from
   the standard input in tab-delimited format as
   follows:

       <host_name>[tab]<svc_description>[tab]<return_code>[tab]<plugin_output>[newline]

   where:

   * `<host_name>` is the short name of host that the service is associated with
   * `<svc_description>` is the description of the service
   * `<return_code>` is the numeric return code
   * `<plugin_output>` is the output from service check

   Host check information is submitted in a similiar
   fashion - just leave out the service description:

       <host_name>[tab]<return_code>[tab]<plugin_output>[newline]



Compiling
---------

The code is very basic and may not work on your particular
system without some tweaking. Most users should be able to compile
the daemon and client piece with the following commands...

    ./configure
    make all

The binaries will be located in the `src/` directory after you
run `make all` and will have to be installed manually.



Installing
----------

The `send_nsca` program and associate config file (`nsca.cfg`) should
be placed on remote machines that you want to have communicate 
with the nsca daemon. This means that you may have to compile the
`send_nsca` program on the remote machine, if its not the same
OS/architecture as that of the central server.

The `nsca` daemon and the configuration file (`nsca.cfg`) should
be placed somewhere on the central server running Nagios.

Notes:

Make sure that you specify and use the same password in
both the `nsca.cfg` and `send_nsca.cfg` files!  If you use a
different password to encrypt the data than you do to decrypt
it, the nsca daemon will reject the data you send it.



Security
--------

There are some security implications with allowing remote clients
to provide service check results to Nagios.  Because of this, you
have the option of encrypting the packets that the NSCA client sends
to the NSCA daemon. Read the `SECURITY` file for more information on
the security risks of running NSCA, along with an explanation of what
kind of protection the encryption provides you.



Running Under `inetd` or `xinetd`
-----------------------------

If you plan on running nsca under `inetd` or `xinetd` and making use
of TCP wrappers, you need to do the following things:

1. Add a line to your `/etc/services` file as follows (modify the port
   number as you see fit)

    nsca            5667/tcp  # NSCA

2. Add entries for the NSCA daemon to either your `inetd` or `xinetd`
   configuration files. Which one you use will depend on which
   superserver is installed on your system. Both methods are described
   below.

   Notes:

   If you run nsca under `inetd` or `xinetd`, the `server_port`
   and `allowed_hosts` variables in the nrpe configuration file are
   ignored.


   * `inetd`
   
     If your system uses the `inetd` superserver **WITH** tcpwrappers, add an
     entry to `/etc/inetd.conf` as follows:

         nsca   stream   tcp   nowait   <user>  /usr/sbin/tcpd  <nscabin> -c <nscacfg> --inetd

     If your system uses the `inetd` superserver **WITHOUT** tcpwrappers, add an
     entry to `/etc/inetd.conf` as follows:

         nsca   stream   tcp   nowait   <user>                  <nscabin> -c <nscacfg> --inetd

      * Replace `<user>` with the name of the user that nsca server should run as (example: `nagios`).
      * Replace `<nscabin>` with the path to the nsca binary on your system (example: `/usr/local/nagios/nsca`).
      * Replace `<nscacfg>` with the path to the nsca config file on your system (example: `/usr/local/nagios/nsca.cfg`).


   * `xinetd`

     If your system uses `xinetd` instead of `inetd`, you'll probably
     want to create a file called `nsca` in your `/etc/xinetd.d`
     directory that contains the following entries (a sample config
     file called `nsca.xinetd` should be created in the root folder of
     the distribution after you run the configure script):

         # default: on
         # description: NSCA
         service nsca
         {
           flags           = REUSE
           socket_type     = stream        
           wait            = no
           user            = <user>
           group           = <group>
           server          = <nscabin>
           server_args     = -c <nscacfg> --inetd
           log_on_failure  += USERID
           disable         = no
           only_from       = <ipaddress1> <ipaddress2> ...
         }


      * Replace `<user>` with the name of the user that the nsca server should run as (example: `nagios`).
      * Replace `<group>` with the name of the group that the nsca server should run as (example: `nagios`).
      * Replace `<nscabin>` with the path to the nsca binary on your system (example: `/usr/local/nagios/nsca`).
      * Replace `<nscacfg>` with the path to the nsca config file on your system (example: `/usr/local/nagios/nsca.cfg`).
      * Replace the `<ipaddress>` fields with the IP addresses of hosts which
        are allowed to connect to the NSCA daemon.  This only works if `xinetd` was
        compiled with support for tcpwrappers.

3. Restart `inetd` or `xinetd`. You know your system best, but some examples are:

       service inetd restart
       service xinetd restart

       systemctl restart inetd.service
       systemctl restart xinetd.service

4. Add entries to your `/etc/hosts.allow` and `/etc/hosts.deny`
   files to enable TCP wrapper protection for the nsca service.
   Alternatively, you can adjust your firewall settings as needed.
   This is an optional step, **although highly recommended**.

Questions?
----------

If you have questions about this addon, or encounter problems getting things
working along the way, your best bet for an answer or quick resolution is to check the
[Nagios Support Forums](https://support.nagios.com/forum/viewforum.php?f=5).


Changelog
**************
NSCA Changelog
**************

x.x.x - xxxx-xx-xx
------------------
- Changed release date to ISO format (yyyy-mm-dd) (John Frickson)
- typofixes (Lajos Veres)
- Send errors to stderr, where they belong (Bas Couwenberg / Sean Finney)
- Fix spelling errors (Bas Couwenberg)
- Fix potential buffer overflow (Bas Couwenberg)
- Spelling fixes (Josh Soref)
- Removed escape newlines so that long output works (Bryan Heden)


2.9.2 - 2016-12-07
------------------
- Renamed configure.in to configure.ac and added check for sigaction (John Frickson)
- Replaced all instances of signal() with sigaction() + blocking (John Frickson)
- Race condition when opening command file (mvela / John Frickson)
- Fix missing argument in open calls (Xavier Bachelot / John Frickson)
- NSCA close/POLLNVAL/accept bug causes hang (mib / John Frickson)


2.9.1 - 01/27/2012
------------------
- Applied patch to allow packets arriving with a future time stamp (Daniel Wittenberg)
- Updated server (nsca) to allow packets with older, smaller packet size (Eric Stanley)


2.9 - 11/04/2011
----------------
- Add config directive to submit directly to checkresults directory (Mike Lindsey)
- Support multi-line check output with 4000 character limit (Mike Lindsey)


2.8 - ??/??/????
----------------
- Added --with-log-facility option to control syslog logging (Ryan Ordway and Brian Seklecki)
- Fixed bug where daemon would segfault if mcrypt library was not initialized before timeout (Holger Weiss)
- Fixed bug with switching from dump file to command file when running under single mode (Ton Voon)
- Fix for small memory leak with running as a single process daemon (Daniel)


2.7.2 - 07/03/2007
------------------
- Fixed bug with NSCA daemon eating CPU if child process couldn't accept a connection in multi-process mode (Chris Wilson)


2.7.1 - 01/29/2007
------------------
- Fixed bug that prevented single mode daemon from working properly
- Added sample scripts for testing functionality to nsca_tests/ (Ton Voon/Altinity)


2.7 - 12/13/2006
----------------
- Fixed crash from malformed command line
- Updated to config.sub and config.guess to latest from GNU Savannah


2.6 - 04/06/2006
----------------
- Spec file fix
- Segfault fix in encryption library cleanup
- Daemon now exits with an error if it can't drop privileges
- Added chroot support (Sean Finney)
- Added support for writing a PID file
- Added support for reloading config files with SIGHUP


2.5 - 01/21/2006
----------------
- Native TCP wrapper support in daemon mode
- Memory leak fix and cleanup (Mark Ferlatte)
- Compiler warning fix (David Binderman)
- max_packet_age=0 will disable packet timestamp checks
- New config.sub and config.guess for checking local system type


2.4 - 07/23/2003
----------------
- Better support for u_int32_t detection
- Minor bug fixes


2.3 - 01/26/2003
----------------
- Minor changes to daemon init code
- Minor Makefile fixes


2.2 - 01/08/2002
----------------
- Hopefully better support for Solaris (u_int32_t)
- Syntax changes for command line arguments
- Added support for passive host checks (supported
  only in Nagios 2.x and later)
- Added sample xinetd config file (nsca.xinetd)
- Minor mods and bug fixes


2.1 - 06/10/2002
----------------
- Fix for compiling under Solaris (Scott Cokely)
- Added setuid/setgid option to config file


2.0 - 02/21/2002
----------------
- Name and version change


1.2.0 - 02/12/2002
------------------
- Compatibility between older versions has been broken!
- Server now sends client IV and timestamp, fixing
  encryption problems and making it more resistant
  to "replay" attacks (Ralf Ertzinger)
- Random IV is now seeded from /dev/urandom instead
  of time() if possible
- Added directions on running under xinetd to README
- Implemented Beej's sendall() to handle incomplete send()s
- Added single-mode daemon code (David Luyer)
- Fixed problem with aggregated writes (David Luyer)
- Better autoconf detection of libmcrypt (Jay McCarthy)
- Added ability to dump check results to alternate
  dump file if command file does not exist (i.e. NetSaint
  is not running)
- Removed some old crud


1.1.2 - 09/22/2001
------------------
- Added append_to_file option to allow for opening the
  command file in either write or append mode
- Explicit integer size in packet structure
- NULL IV is used for both encryption/decryption.  This
  leaves the daemon open to "replay" attacks, but will
  disappear in future revisions when IV negotiation
  between the client and daemon is implemented.


1.1.1 - 04/26/2001
------------------
- Configure script bug fix


1.1.0 - 02/24/2001
------------------
- Added encryption routines (using libmcrypt)
- Minor code enhancements/cleanups


1.0b3 - 12/21/2000
------------------
- Removed lock file code, as 0.0.7 uses a named pipe
- Documented aggregated_writes option


1.0b2 - 08/26/2000
------------------
- Added option to use write lock when updating external
  command file


1.0b1 - 04/17/2000
------------------
- Initial cut


include/common.h
/************************************************************************
 *
 * COMMON.H - NSCA Common Include File
 * Copyright (c) 1999-2003 Ethan Galstad (nagios@nagios.org)
 * Last Modified: 12-07-2016
 *
 * License:
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 ************************************************************************/

#include "config.h"


#define PROGRAM_VERSION "2.9.2"
#define MODIFICATION_DATE "12-07-2016"


#define OK		0
#define ERROR		-1

#define TRUE		1
#define FALSE		0

#define STATE_UNKNOWN  	3	/* service state return codes */
#define	STATE_CRITICAL 	2
#define STATE_WARNING 	1
#define STATE_OK       	0

#define DEFAULT_SOCKET_TIMEOUT	10	/* timeout after 10 seconds */

#define MAX_INPUT_BUFFER	5120	/* max size of most buffers we use */

#define MAX_HOST_ADDRESS_LENGTH	256	/* max size of a host address */

/**************************************************************************************/
/* WARNING!                                                                           */
/*                                                                                    */
/* Changing the lengths below may cause packet failures between clients and servers   */
/* of different versions.                                                             */
/**************************************************************************************/

#define MAX_HOSTNAME_LENGTH	64
#define MAX_DESCRIPTION_LENGTH	128
#define MAX_PLUGINOUTPUT_LENGTH	4096

#define OLD_PLUGINOUTPUT_LENGTH	512
#define OLD_PACKET_LENGTH (( sizeof( data_packet) - ( MAX_PLUGINOUTPUT_LENGTH - OLD_PLUGINOUTPUT_LENGTH)))

#define MAX_PASSWORD_LENGTH     512

#define BLOCK_DELIMITER  "\x17"


/********************* ENCRYPTION TYPES ****************/

#define ENCRYPT_NONE            0       /* no encryption */
#define ENCRYPT_XOR             1       /* not really encrypted, just obfuscated */

#ifdef HAVE_LIBMCRYPT
#define ENCRYPT_DES             2       /* DES */
#define ENCRYPT_3DES            3       /* 3DES or Triple DES */
#define ENCRYPT_CAST128         4       /* CAST-128 */
#define ENCRYPT_CAST256         5       /* CAST-256 */
#define ENCRYPT_XTEA            6       /* xTEA */
#define ENCRYPT_3WAY            7       /* 3-WAY */
#define ENCRYPT_BLOWFISH        8       /* SKIPJACK */
#define ENCRYPT_TWOFISH         9       /* TWOFISH */
#define ENCRYPT_LOKI97          10      /* LOKI97 */
#define ENCRYPT_RC2             11      /* RC2 */
#define ENCRYPT_ARCFOUR         12      /* RC4 */
#define ENCRYPT_RC6             13      /* RC6 */            /* UNUSED */
#define ENCRYPT_RIJNDAEL128     14      /* RIJNDAEL-128 */
#define ENCRYPT_RIJNDAEL192     15      /* RIJNDAEL-192 */
#define ENCRYPT_RIJNDAEL256     16      /* RIJNDAEL-256 */
#define ENCRYPT_MARS            17      /* MARS */           /* UNUSED */
#define ENCRYPT_PANAMA          18      /* PANAMA */         /* UNUSED */
#define ENCRYPT_WAKE            19      /* WAKE */
#define ENCRYPT_SERPENT         20      /* SERPENT */
#define ENCRYPT_IDEA            21      /* IDEA */           /* UNUSED */
#define ENCRYPT_ENIGMA          22      /* ENIGMA (Unix crypt) */
#define ENCRYPT_GOST            23      /* GOST */
#define ENCRYPT_SAFER64         24      /* SAFER-sk64 */
#define ENCRYPT_SAFER128        25      /* SAFER-sk128 */
#define ENCRYPT_SAFERPLUS       26      /* SAFER+ */
#endif



/******************** MISC DEFINITIONS *****************/

#define TRANSMITTED_IV_SIZE     128     /* size of IV to transmit - must be as big as largest IV needed for any crypto algorithm */


/*************** PACKET STRUCTURE DEFINITIONS **********/

#define NSCA_PACKET_VERSION_3   3		/* packet version identifier */
#define NSCA_PACKET_VERSION_2	2		/* older packet version identifiers */
#define NSCA_PACKET_VERSION_1	1

/* data packet containing service check results */
typedef struct data_packet_struct{
	int16_t   packet_version;
	u_int32_t crc32_value;
	u_int32_t timestamp;
	int16_t   return_code;
	char      host_name[MAX_HOSTNAME_LENGTH];
	char      svc_description[MAX_DESCRIPTION_LENGTH];
	char      plugin_output[MAX_PLUGINOUTPUT_LENGTH];
        }data_packet;

/* initialization packet containing IV and timestamp */
typedef struct init_packet_struct{
	char      iv[TRANSMITTED_IV_SIZE];
	u_int32_t timestamp;
        }init_packet;




/**************** OPERATING SYSTEM SPECIFIC DEFINITIONS **********/
#ifdef __sun

#  ifndef LOG_AUTHPRIV
#    define LOG_AUTHPRIV LOG_AUTH
#  endif

#  ifndef LOG_FTP
#    define LOG_FTP LOG_DAEMON
#  endif

#endif
